### PR TITLE
Allow a low memory option for index construction

### PR DESCRIPTION
**Pull request by lmcinnes**
This splits out the internals of of the NNdescent loop to either use or skip the use of the ``tried`` cache depending on whether the user selects a ``low_memory`` option which defaults to ``False``.

### CODE REVIEW

File: pynndescent/sparse_nndescent.py [hunk 1]
```diff
@@ -70,43 +71,84 @@ def sparse_nn_descent(
...
+    n_vertices,
+    n_neighbors,
+    rng_state,
+    max_candidates=50,
+    sparse_dist=sparse_euclidean,
+    dist_args=(),
+    n_iters=10,
+    delta=0.001,
+    rho=0.5,
+    tried=set([(-1, -1)]),
```
Comment (jlmelville) [inline pynndescent/sparse_nndescent.py]: Same comment as for the non-sparse case, but just to be on the safe side, repeating myself:

Is this safe? The default `tried` will only be initialized once, so the same set will be reused if `sparse_nn_descent_internal_high_memory` is called multiple times.

File: pynndescent/pynndescent_.py [hunk 2]
```diff
@@ -167,28 +168,72 @@ def nn_descent(
...
+    data,
+    n_neighbors,
+    rng_state,
+    max_candidates=50,
+    dist=dist.euclidean,
+    dist_args=(),
+    n_iters=10,
+    delta=0.001,
+    rho=0.5,
+    tried=set([(-1, -1)]),
```
Comment (jlmelville) [inline pynndescent/pynndescent_.py]: Is this safe? The default `tried` will only be initialized once, so the same set will be reused if `nn_descent_internal_high_memory` is called multiple times.

Comment (lmcinnes) [reply pynndescent/pynndescent_.py]: Good point -- the catch is that it needs an instantiation like this to get numba to infer the type. Ultimately this is an internal routine and should not get called without supplying tried, so perhaps the better solution is to make it a required argument.

**Comment by lmcinnes** [comment]
Thanks for the suggestions!

**Pull request merged**